src/main/java/com/cloudbees/plugins/credentials/impl/BaseStandardCredentials.java

```diff
@@ -109,4 +117,35 @@ public final boolean equals(Object o) {
     public final int hashCode() {
         return IdCredentials.Helpers.hashCode(this);
     }
+
+    /**
+     * Descriptor to use for subclasses of {@link BaseStandardCredentials}.
+     * <p>{@code <st:include page="id-and-description" class="${descriptor.clazz}"/>} in {@code credentials.jelly} to pick up standard controls for {@link #getId} and {@link #getDescription}.
+     */
+    protected static abstract class BaseStandardCredentialsDescriptor extends CredentialsDescriptor {
+
+        protected BaseStandardCredentialsDescriptor() {
+            clazz.asSubclass(BaseStandardCredentials.class);
+        }
+
+        protected BaseStandardCredentialsDescriptor(Class<? extends BaseStandardCredentials> clazz) {
+            super(clazz);
+        }
+
+        public final FormValidation doCheckId(@QueryParameter String value) {
+            if (value.isEmpty()) {
+                return FormValidation.ok();
+            }
+            if (!value.matches("[a-zA-Z0-9_.-]+")) { // anything else considered kosher?
+                return FormValidation.error("Unacceptable characters");
+            }
+            // TODO perhaps limit to the store associated with a User or ItemGroup (Jenkins/Folder) in @AncestorInPath
+            if (CredentialsMatchers.firstOrNull(CredentialsProvider.lookupCredentials(IdCredentials.class, (ItemGroup<?>) null, null, Collections.<DomainRequirement>emptyList()), CredentialsMatchers.withId(value)) != null) {
```

Comment:
> may want to check current user's credentials

Probably right.

> should be limited to the store that this credential is being created within […]
> a duplicate in the current store should be an ERROR. a duplicate globally should be a WARNING

Agreed, this would be ideal. I will probably need help with the API here. I know how to get a `User` or `ItemGroup` from the request, but am not sure how to select the corresponding `CredentialsStore` (especially when inheritance of credentials from parent folders/root is considered).